Lib/test/test_poplib.py

```diff
@@ -306,9 +307,29 @@ def test_noop(self):
     def test_rpop(self):
         self.assertOK(self.client.rpop('foo'))
 
-    def test_apop(self):
+    def test_apop_normal(self):
         self.assertOK(self.client.apop('foo', 'dummypassword'))
 
+    def test_apop_REDOS(self):
+        old_welcome = self.client.welcome
+
+        # Replace welcome with very long evil welcome.
+        # NB The upper bound on welcome length is currently 2048.
+        # At this length, evil input makes each apop call take
+        # on the order of milliseconds instead of microseconds.
+        evil_welcome = bytes('+OK' + ('<' * 500000), self.client.encoding)
+        self.client.welcome = evil_welcome
+
+        threw = False
+        try:
+            self.client.apop('foo', 'dummypassword')
+        except poplib.error_proto:
+            threw = True
+        # The evil welcome is invalid, so apop should throw.
+        self.assertTrue(threw)
```

Comment:
Why not use `assertRaises()`?

Reply:
The reason for the poor choices of asserts is that I didn't know what asserts are available and made do with what I could find. I really appreciate your help.

Reply:
https://docs.python.org/3/library/unittest.html#test-cases

`assertRaises()` can be used as a function:
```python
self.assertRaises(poplib.error_proto, self.client.apop, 'foo', 'dummypassword')
```
or as a context manager:
```python
with self.assertRaises(poplib.error_proto):
    self.client.apop('foo', 'dummypassword')
```

Reply:
Yup I have at last found the documentation for unittest.